modules/core/src/main/java/org/apache/ignite/internal/processors/platform/utils/PlatformUtils.java

```diff
@@ -833,12 +833,26 @@ else if (err instanceof IgniteException)
      */
     public static Object readInvocationResult(PlatformContext ctx, BinaryRawReaderEx reader)
         throws IgniteCheckedException {
+        return readInvocationResult(ctx, reader, false);
+    }
+
+    /**
+     * Reads invocation result (of a job/service/etc) using a common protocol.
+     *
+     * @param ctx Platform context.
+     * @param reader Reader.
+     * @param deserialize If {@code true} deserialize invocation result.
+     * @return Result.
+     * @throws IgniteCheckedException When invocation result is an error.
+     */
+    public static Object readInvocationResult(PlatformContext ctx, BinaryRawReaderEx reader, boolean deserialize)
+            throws IgniteCheckedException {
         // 1. Read success flag.
         boolean success = reader.readBoolean();
 
         if (success)
             // 2. Return result as is.
-            return reader.readObjectDetached();
+            return deserialize ? reader.readObject() : reader.readObjectDetached();
```

Comment:
I think it's not enough for correct deserialization sometimes (for arrays and collection of objects). We should also do `unwrapBinary()`.
Also (optionally), we can use `unwrapBinary()` directly after method invocation in `GridServiceProxy` instead of adding new methods to `PlatformUtils`, `PlatformService`, `PlatformAbstractService`.

Reply:
Unfortunatelly, the approach you suggested doesn't work, but is more elegant, without doubts. But these methods returns Object and need more additional works. But my approach works and tests prove this. So I suggest leave it as is.